File: internal/utils/bitstream.go
package utils

type bitstreamReader struct {
	bytes    []int
	bytesize int
	pos      int
	bits     int
	value    int
}

// NewBitstreamReader returns a new instance of the bitstreamReader struct
func NewBitstreamReader(bytes []int, bytesize int) *bitstreamReader {
	reader := &bitstreamReader{}
	reader.bytes = bytes
	reader.bytesize = bytesize
	reader.pos = 0
	reader.bits = 0
	reader.value = 0
	return reader
}

// Remaining returns a boolean stating whether there is still
// something to read in the bitstream
func (br *bitstreamReader) Remaining() bool {
	if br.pos < len(br.bytes) {
		return true
	}
	if br.bits > 0 {
		return true
	}
	return false
}

// Read reads a given number of bits from the bitstream
func (br *bitstreamReader) Read(count int) int {
	for br.bits < count {
		if br.pos >= len(br.bytes) {
			break
		}
		br.value |= (br.bytes[br.pos] & ((1 << br.bytesize) - 1)) << br.bits
		br.bits += br.bytesize
		br.pos++
	}

	ret := br.value & ((1 << count) - 1)
	br.value >>= count
	br.bits -= count
	return ret
}

type bitstreamWriter struct {
	bytes    []int
	bytesize int
	bits     int
	value    int
}

// NewBitstreamWriter returns a new instance of the bitstreamWriter struct
func NewBitstreamWriter(bytesize int) *bitstreamWriter {
	writer := &bitstreamWriter{}
	writer.bytes = make([]int, 0)
	writer.bytesize = bytesize
	writer.bits = 0
	writer.value = 0
	return writer
}

// Finish returns the writen bitstream
func (bw *bitstreamWriter) Finish() []int {
	if bw.bits > 0 {
		bw.bytes = append(bw.bytes, bw.value&((1<<bw.bytesize)-1))
	}
	return bw.bytes
}

// Write writes the supplied value into the bitstream
// with the bit length being specified
func (bw *bitstreamWriter) Write(value int, bits int) {
	bw.value |= (value & ((1 << bits) - 1)) << bw.bits
	bw.bits += bits
	for bw.bits >= bw.bytesize {
		bw.bytes = append(bw.bytes, bw.value&((1<<bw.bytesize)-1))
		bw.value >>= bw.bytesize
		bw.bits -= bw.bytesize
	}
}

File: src/lib/bitstream.ts
/**
 * based off
 * https://github.com/mid-kid/pmdrtdx_passwords/blob/master/password.py#L44
 */
export class BitstreamReader {
    bytes: bigint[];
    bytesize: bigint;
    pos: number;
    bits: bigint;
    value: bigint;
    constructor(bytes: number[], bytesize: number) {
        this.bytes = [];
        for (const byte of bytes) {
            this.bytes.push(BigInt(byte));
        }
        this.bytesize = BigInt(bytesize);
        this.pos = 0;
        this.bits = 0n;
        this.value = 0n;
    }

    remaining(): boolean {
        if (this.pos < this.bytes.length) return true;
        if (this.bits > 0n) return true;
        return false;
    }

    read(count: number | bigint): number {
        if (typeof count === 'number') count = BigInt(count);
        while (this.bits < count) {
            if (this.pos >= this.bytes.length) break;
            this.value |= (this.bytes[this.pos] & ((1n << this.bytesize) - 1n)) << this.bits;
            this.bits += this.bytesize;
            this.pos++;
        }
        const ret = this.value & ((1n << count) - 1n);
        this.value >>= count;
        this.bits -= count;
        return Number(ret);
    }
}

/**
 * based off
 * https://github.com/mid-kid/pmdrtdx_passwords/blob/master/password.py#L72
 */
export class BitstreamWriter {
    bytes: number[];
    bytesize: bigint;
    bits: bigint;
    value: bigint;
    constructor(bytesize: number) {
        this.bytes = [];
        this.bytesize = BigInt(bytesize);
        this.bits = 0n;
        this.value = 0n;
    }

    finish(): number[] {
        if (this.bits > 0) this.bytes.push(Number(this.value & ((1n << this.bytesize) - 1n)));
        return this.bytes;
    }

    write(value: number | bigint, bits: number | bigint): void {
        if (typeof value === 'number') value = BigInt(value);
        if (typeof bits === 'number') bits = BigInt(bits);
        this.value |= (value & ((1n << bits) - 1n)) << this.bits;
        this.bits += bits;
        while (this.bits >= this.bytesize) {
            this.bytes.push(Number(this.value & ((1n << this.bytesize) - 1n)));
            this.value >>= this.bytesize;
            this.bits -= this.bytesize;
        }
    }
}
